### PR TITLE
[jit] Split out Function into its own file

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21539 [jit] Split out Function into its own file**

Some preparatory movement to prepare for the CU ownership overhaul.
Changes are purely mechanical.

Differential Revision: [D15741979](https://our.internmc.facebook.com/intern/diff/D15741979)

### CODE REVIEW

**Review by zdevito — APPROVED** [review]

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1093,7 +1093,7 @@ using FunctionTypePtr = std::shared_ptr<FunctionType>;
 // This type represents a Python Function
 struct CAFFE2_API FunctionType : public Type {
   static FunctionTypePtr create(
-      std::shared_ptr<torch::jit::script::Function> function) {
+      std::shared_ptr<torch::jit::Function> function) {
```
Comment (Krovatkin) [inline aten/src/ATen/core/jit_type.h]: flatenning namespaces?

File: torch/csrc/jit/function.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  // that it can construct methods out of order
+  std::function<void(Function&)> function_creator_;
+
+  // if absent, then we generate a default schema based on the graph
+  // mutable because getSchema caches the default schema if one is requested
+  // before a call to setSchema
+  mutable std::unique_ptr<FunctionSchema> schema_;
+};
+} // namespace jit
+} // namespace torch
```
Comment (Krovatkin) [inline torch/csrc/jit/function.h]: nitpick: newline

File: torch/csrc/jit/function.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    throw script::ErrorReport() // TODO: once lower_first_class methods is
+                                // removed re-establish callsite info for
+                                // debugging
+        << " method '" << name() << "' is called recursively. "
+        << "Recursive calls are not supported";
+  }
+  check_single_output();
+}
+} // namespace jit
+} // namespace torch
```
Comment (Krovatkin) [inline torch/csrc/jit/function.cpp]: nitpick: newline

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: aten/src/ATen/core/jit_type.h (hunk 1 above)
Comment (suo) [reply aten/src/ATen/core/jit_type.h]: yes, thought I might as well start now :)

**Pull request closed without merging**